File: bindings/java/com/statespec/backend/Observability.java
package com.statespec.backend;

import com.statespec.backend.Backend.BackendException;
import com.statespec.backend.Backend.Transaction;
import java.util.Collections;
import java.util.Map;
import java.util.TreeMap;

public interface Observability
{
    enum LogLevel
    {
        DEBUG,
        INFO,
        WARN,
        ERROR
    }

    enum MetricKind
    {
        COUNTER,
        GAUGE,
        HISTOGRAM
    }

    record LogEvent(
        String name,
        LogLevel level,
        String eventName,
        Map<String,
            Json> fields
    )
    {
        public LogEvent
        {
            fields = Collections.unmodifiableMap(new TreeMap<>(fields));
        }
    }

    record MetricSample(
        String name,
        MetricKind kind,
        String backendName,
        double value,
        String unit,
        Map<String,
            Json> labels
    )
    {
        public MetricSample
        {
            labels = Collections.unmodifiableMap(new TreeMap<>(labels));
        }
    }

    void emitLog(
        Backend backend,
        LogEvent event
    ) throws BackendException;

    void emitLogTx(
        Transaction tx,
        LogEvent event
    ) throws BackendException;

    void recordMetric(
        Backend backend,
        MetricSample sample
    ) throws BackendException;

    void recordMetricTx(
        Transaction tx,
        MetricSample sample
    ) throws BackendException;
}

File: bindings/rust/observability.rs
use std::collections::BTreeMap;

use crate::backend::{Backend, BackendResult};
use crate::json::Json;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MetricKind {
    Counter,
    Gauge,
    Histogram,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LogEvent {
    pub name: String,
    pub level: LogLevel,
    pub event_name: String,
    pub fields: BTreeMap<String, Json>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MetricSample {
    pub name: String,
    pub kind: MetricKind,
    pub backend_name: String,
    pub value: f64,
    pub unit: String,
    pub labels: BTreeMap<String, Json>,
}

pub trait ObservabilitySink<B: Backend> {
    fn emit_log(&self, backend: &B, event: &LogEvent) -> BackendResult<()>;

    fn emit_log_tx(&self, tx: &mut B::Tx, event: &LogEvent) -> BackendResult<()>;

    fn record_metric(&self, backend: &B, sample: &MetricSample) -> BackendResult<()>;

    fn record_metric_tx(&self, tx: &mut B::Tx, sample: &MetricSample) -> BackendResult<()>;
}
